src/Microsoft.Azure.SignalR.Common/Endpoints/ServiceEndpointManagerBase.cs

```diff
@@ -96,15 +100,199 @@ protected ServiceEndpoint[] GetValuableEndpoints(IEnumerable<ServiceEndpoint> en
             return groupedEndpoints.ToArray();
         }
 
+        protected async Task AddServiceEndpointsAsync(IReadOnlyList<ServiceEndpoint> endpoints, TimeSpan timeout)
+        {
+            if (endpoints.Count > 0)
+            {
+                try
+                {
+                    var hubEndpoints = CreateHubServiceEndpoints(endpoints, true);
+
+                    await Task.WhenAll(hubEndpoints.Select(e => AddHubServiceEndpointAsync(e, timeout)));
+
+                    // TODO: update local store for negotiation
+                }
+                catch
+                {
+                    Log.FailedAddingEndpoints(_logger);
+                }
+            }
+        }
+
+        protected async Task RemoveServiceEndpointsAsync(IReadOnlyList<ServiceEndpoint> endpoints, TimeSpan timeout)
+        {
+            if (endpoints.Count > 0)
+            {
+                try
+                {
+                    var hubEndpoints = CreateHubServiceEndpoints(endpoints, true);
+
+                    // TODO: update local store for negotiation
+
+                    await Task.WhenAll(hubEndpoints.Select(e => RemoveHubServiceEndpointAsync(e, timeout)));
```

Comment:
why not timeout here but inside every single task?

Reply:
Cause it's originally planned for single endpoint scale. Considering parallel do task, I've added timeout check in `ServiceEndpointManager` side which could reduce by pass timeout all the way, and use the CancellationToken inside to clean the task and ensure process won't be blocked.